### PR TITLE
SLS Enhancements

### PR DESCRIPTION
**Pull request by raghits**
### Requirements

* Refine the working of handling templates
* Support for subscription log plugin
* Support for lambda edge plugin

### Description of the Change

* Changes made to simplify the process.

### Benefits

* Improved user experiences.
* Simplified Configuration

### Possible Drawbacks

None.

### Applicable Issues

None.

### CODE REVIEW

File: builds/jazz-build-module/sls-app/sbr.groovy [hunk 1]
```diff
@@ -1120,18 +1123,38 @@ def prepareServerlessYml(aConfig, env, configLoader, envDeploymenDescriptor) {
...
     }
+  } catch(e) { // TODO to catch the type error
+      echo "The user supplied serverless.yml does not exist in the code. So the default value from config will be used: ${e}"
+  }
 
     def doc = deploymentDescriptor  ? readYaml(text: deploymentDescriptor ) : [:] // If no descriptor present then simply making an empty one. The readYaml default behavior is to return empty string back that is harful as Map not String is expected below
 
+    def envKinesisStreamArn = null
+    
+    switch (env)
```
Comment (devsatishm) [inline builds/jazz-build-module/sls-app/sbr.groovy]: This is incorrect. We always use `configLoader.JAZZ.PLATFORM.AWS.KINESIS_LOGS_STREAM.PROD` for user services.

`configLoader.JAZZ.PLATFORM.AWS.KINESIS_LOGS_STREAM.STG` & `configLoader.JAZZ.PLATFORM.AWS.KINESIS_LOGS_STREAM.DEV` is for developing services that read messages off of the stream like dev versions of log-streamer services.

File: builds/jazz-build-module/sls-app/sbr.groovy [hunk 2]
```diff
@@ -1120,18 +1123,38 @@ def prepareServerlessYml(aConfig, env, configLoader, envDeploymenDescriptor) {
...
+        break;
+      default: 
+        envKinesisStreamArn = configLoader.JAZZ.PLATFORM.AWS.KINESIS_LOGS_STREAM.DEV;
+        break;
+    }
+
     context =["environment_logical_id": env,
             "INSTANCE_PREFIX": configLoader.INSTANCE_PREFIX,
             "REGION": aConfig.region,
             "cloud_provider": "aws",
```
Comment (devsatishm) [inline builds/jazz-build-module/sls-app/sbr.groovy]: cloud_provider => cloudProvider

File: builds/jazz-build-module/sls-app/sbr.groovy [hunk 3]
```diff
@@ -1120,18 +1123,38 @@ def prepareServerlessYml(aConfig, env, configLoader, envDeploymenDescriptor) {
...
+            "kinesisStreamArn": envKinesisStreamArn,
+            "platformRoleArn": configLoader.AWS.ACCOUNTS.find{ it.ACCOUNTID == aConfig.accountId}.IAM.PLATFORMSERVICES_ROLEID, // pick the role for selected account
             "serverless_framework_version": ">=1.0.0 <2.0.0"]
```
Comment (devsatishm) [inline builds/jazz-build-module/sls-app/sbr.groovy]: `serverlessFrameworkVersion`

File: core/jazz_assets/swagger/swagger.json [hunk 4]
```diff
@@ -1145,29 +1185,139 @@
...
+        "cloudwatch_event",
+        "swagger_url",
+        "endpoint_url",
+        "apigee_proxy",
+        "dynamodb_stream",
+        "iam_role",
+        "sqs",
+        "kinesis_stream",
+        "sequence_diagram_url",
+        "certificate",
```
Comment (devsatishm) [inline core/jazz_assets/swagger/swagger.json]: let's remove the unsupported asset types - sequence_diagram_url, certificate & dns_record

File: core/jazz_logs/swagger/swagger.json [hunk 5]
```diff
@@ -479,8 +479,12 @@
             "apigee_proxy"
```
Comment (devsatishm) [inline core/jazz_logs/swagger/swagger.json]: This list is probably incorrect. We only support logs for `lambda` & `apigateway`.

File: builds/jazz-build-module/sls-app/sbr.groovy (hunk 1 above)
Comment (raghits) [reply builds/jazz-build-module/sls-app/sbr.groovy]: done

File: builds/jazz-build-module/sls-app/sbr.groovy (hunk 2 above)
Comment (raghits) [reply builds/jazz-build-module/sls-app/sbr.groovy]: done

File: builds/jazz-build-module/sls-app/sbr.groovy (hunk 3 above)
Comment (raghits) [reply builds/jazz-build-module/sls-app/sbr.groovy]: done

File: core/jazz_assets/swagger/swagger.json (hunk 4 above)
Comment (raghits) [reply core/jazz_assets/swagger/swagger.json]: done

File: core/jazz_logs/swagger/swagger.json (hunk 5 above)
Comment (raghits) [reply core/jazz_logs/swagger/swagger.json]: done

**Comment by raghits** [comment]
@devsatishm Addressed on CR comments. Also included all changes needed to support log subscription and lambda edge plugin.

File: builds/jazz-build-module/sls-app/serverless-build-rules.yml [hunk 6]
```diff
@@ -398,15 +399,21 @@ provider:
   vpc: # Optional VPC. But if you use VPC then both subproperties (securityGroupIds and subnetIds) are required
     securityGroupIds:
       sbr-type: "[aws-id]"
-      sbr-render: user-only
+      sbr-render: config-wins
```
Comment (devsatishm) [inline builds/jazz-build-module/sls-app/serverless-build-rules.yml]: VPC support isn't available yet. `config-wins` should be reverted back to `user-only` & `sbr-mandatory` to false. Same with `subnetIds `

File: builds/jazz-build-module/sls-app/serverless-build-rules.yml [hunk 7]
```diff
@@ -477,15 +484,20 @@ package: # Optional deployment packaging configuration
...
+# works as expected when using sbr-formula only NOT with sbr-example
 plugins:
-    sbr-type: "[aws-var-name:plugin]"
-    sbr-render: user-only
+    sbr-type: "[plugin]"
+    sbr-render: merge
+    sbr-mandatory: true
+    sbr-value:
+      sbr-formula:
+       - serverless-plugin-log-subscription
```
Comment (devsatishm) [inline builds/jazz-build-module/sls-app/serverless-build-rules.yml]: Quick question - What if platform needs to inject multiple plugins by default?

File: core/jazz_assets/swagger/swagger.json (hunk 4 above)
Comment (devsatishm) [reply core/jazz_assets/swagger/swagger.json]: Looks like these are not updated yet.

File: builds/jazz-build-module/sls-app/serverless-build-rules.yml (hunk 7 above)
Comment (raghits) [reply builds/jazz-build-module/sls-app/serverless-build-rules.yml]: Then we would need to provide that list here.

File: builds/jazz-build-module/sls-app/serverless-build-rules.yml (hunk 6 above)
Comment (raghits) [reply builds/jazz-build-module/sls-app/serverless-build-rules.yml]: I guess this applies only to internal. Removed.

File: core/jazz_assets/swagger/swagger.json (hunk 4 above)
Comment (raghits) [reply core/jazz_assets/swagger/swagger.json]: Thanks for catching that. done.

**Comment by devsatishm** [comment]
LGTM

**Pull request merged**